elastic/beats
# [Heartbeat] States reloader and eslegclient

## What does this PR do?

Requires https://github.com/andrewvc/integrations/pull/1.
Supersedes https://github.com/elastic/beats/issues/33357.
Fixes https://github.com/elastic/beats/issues/33354, https://github.com/elastic/beats/issues/33299 and https://github.com/elastic/beats/issues/33244

This PR:
- Introduces a `DeferredStateLoader` reloader for managed output section of heartbeat instances. It holds monitors from running state queries until the agent has had enough time to initialize it async or a default timeout is triggered.
- Replaces the use of `elasticsearch.Client` with `eslegclient.Connection`, which supports all `output` options by default.
- Fixes a bug where state would be retrieved even when found.

## Checklist




## How to test this PR locally



To test `eslegclient`:
1. Checkout this PR and build HB locally.
2. Run ES locally with self-signed certificates and use that as HB default output with disabled ssl verification.
3. Run/ stop HB multiple times. Check in discover that states are reused.

To test output reloader:
1. Build Heartbeat and Elastic-agent locally.
2. Enroll a container in a private location policy multiple times and check that states are reused.

## Screenshots

![image](https://user-images.githubusercontent.com/95703246/197028752-facd9ce4-34d0-4d1e-91f9-6a0410ea1e6c.png)
